### PR TITLE
Improving g3c test robustness

### PR DESCRIPTION
**Pull request by hugohadfield**
~~The goal of this PR will be to improve test robustness, linked to https://github.com/pygae/clifford/issues/309
Aiming to do:
- [ ] Convert npt.assert_almost_equal to npt.assert_allclose
- [ ] Integrate the [hypothesis package](https://hypothesis.readthedocs.io/en/latest/)
- [ ] Look at seeding non deterministic tests
- [ ] Improve test coverage

Feedback and suggestions welcome
~~

Updated PR goals:
Improve g3c test robustness and improve g3c tools, linked to https://github.com/pygae/clifford/issues/309
Convert npt.assert_almost_equal to npt.assert_allclose
Improve test coverage
Fix known broken functions

### CODE REVIEW

File: clifford/test/test_g3c_tools.py [hunk 1]
```diff
@@ -29,14 +29,17 @@
...
+
 @too_slow_without_jit
 class TestRotorGeneration:
     def test_generate_translation_rotor(self):
         for i in range(10000):
             euc_vector_a = random_euc_mv()
             res = generate_translation_rotor(euc_vector_a)
             res2 = (1 + ninf * euc_vector_a / 2)
-            npt.assert_almost_equal(res.value, res2.value)
+            npt.assert_allclose(res.value, res2.value, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT)
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:42]: You might want to use, at the top of this file,
```
assert_allclose = functions.partial(npt.assert_allclose, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT)
```
To save yourself some typing

Comment (hugohadfield) [reply clifford/test/test_g3c_tools.py:42]: *functools :)

**Comment by hugohadfield** [comment]
closing and reopening has kicked off travis :champagne:

File: clifford/test/test_g3c_tools.py [hunk 2]
```diff
@@ -247,8 +252,9 @@ def test_dual(self):
         up_array = test_array.up()
         I5 = layout.blades['e12345']
 
-        npt.assert_almost_equal((up_array * ConformalMVArray([I5])).value,
-                                ConformalMVArray([i * I5 for i in up_array]).value)
+        npt.assert_allclose((up_array * ConformalMVArray([I5])).value,
+                ConformalMVArray([i * I5 for i in up_array]).value,
+                            rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT)
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:257]: flake8 is failing due to the weird indent.
```suggestion
        npt.assert_allclose(
                (up_array * ConformalMVArray([I5])).value,
                ConformalMVArray([i * I5 for i in up_array]).value,
                rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT)
```

File: clifford/test/test_g3c_tools.py [hunk 3]
```diff
@@ -334,11 +371,85 @@ def test_enclosing_spheres(self):
...
+            new_point = new_point/len(lines)
+            assert_allclose(new_point.value, res.value)
+
+    def test_get_truncated_line_reflection_matrix(self):
+        for i in range(10):
+            lines = [random_line() for i in range(10)]
+            Lmat = get_line_reflection_matrix(lines, 1)
+            Lmat_trunc = val_truncated_get_line_reflection_matrix(np.array([l.value for l in lines]), 1)
+            assert_allclose(Lmat_trunc, Lmat[1:6, 1:6])
+
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:422]: flake8 failure is caused by
```suggestion

```

File: clifford/test/test_g3c_tools.py [hunk 4]
```diff
@@ -420,7 +420,6 @@ def test_get_truncated_line_reflection_matrix(self):
             Lmat = get_line_reflection_matrix(lines, 1)
             Lmat_trunc = val_truncated_get_line_reflection_matrix(np.array([l.value for l in lines]), 1)
             assert_allclose(Lmat_trunc, Lmat[1:6, 1:6])
-            
     def test_get_midpoint_between_lines(self):
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:423]:
```suggestion

    def test_get_midpoint_between_lines(self):
```

File: clifford/test/test_g3c_tools.py (hunk 3 above)
Comment (eric-wieser) [reply clifford/test/test_g3c_tools.py]: Whoops, needed a line here still, just not with spaces on the end of it.

**Comment by hugohadfield** [comment]
Closed and reopened to kick off CI again etc

File: clifford/test/test_g3c_tools.py [hunk 5]
```diff
@@ -22,21 +22,26 @@
...
+import functools
 
 too_slow_without_jit = pytest.mark.skipif(
     numba.config.DISABLE_JIT, reason="test is too slow without JIT"
 )
 
 
+RTOL_DEFAULT = 1E-4
+ATOL_DEFAULT = 1E-6
+assert_allclose = functools.partial(npt.assert_allclose, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT)
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:34]: There's no point doing this unless you're going to call it below!

File: clifford/test/test_g3c_tools.py [hunk 6]
```diff
@@ -334,11 +372,84 @@ def test_enclosing_spheres(self):
...
+            X1, X2 = iterative_closest_points_circle_line(C, L, niterations=50)
+            X1Andreas = closest_point_on_circle_from_line(C, L)
+            X2Andreas = closest_point_on_line_from_circle(C, L)
+
+            assert_allclose((X1 ^ C).value, 0)
+            assert_allclose((X1Andreas ^ C).value, 0)
+            assert_allclose((X2 ^ L).value, 0)
+            assert_allclose((X2Andreas ^ L).value, 0)
+            print(X1 | X2)
+            print(X1Andreas | X2Andreas)
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:403]: Are these useful?

File: clifford/tools/g3c/__init__.py [hunk 7]
```diff
@@ -428,10 +431,74 @@ def closest_points_on_circles(C1, C2, niterations=20):
...
+def closest_point_on_circle_from_line(C, L):
+    """
+    Returns the point on the circle C that is closest to the line L
+    Uses the algorithm described in Appendix A of Andreas Aristidou's PhD thesis
+    """
+    phi = (C^einf).normal()
+    B = meet(L, phi)
+    A = normalise_n_minus_1((C * einf * C)(1))
+    bound_sphere = ((C * phi) * I5).normal()
+    if abs((B**2)[0]) < 1E-6:
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:451]: can you make this `1e-6`  an `eps` argument to this function?

File: clifford/tools/g3c/__init__.py [hunk 8]
```diff
@@ -428,10 +431,74 @@ def closest_points_on_circles(C1, C2, niterations=20):
...
+            # Line is perpendicular to the plane of the circle
+            L2 = A ^ Y ^ einf
+        else:
+            L2 = P ^ Y ^ einf
+    else:
+        L2 = A ^ Y ^ einf
+    PP = meet(L2, bound_sphere)
+    Xs = point_pair_to_end_points(PP)
+    ind = np.argmax([(Xs[0]|P)[0], (Xs[1]|P)[0]])
+    return Xs[ind]
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:491]: No need to use numpy here
```suggestion
    return max(Xs, key=lambda x: (x|P)[0])
```

File: clifford/tools/g3c/__init__.py [hunk 9]
```diff
@@ -560,14 +625,14 @@ def val_truncated_get_line_reflection_matrix(line_array, n_power):
     mat2solve = np.zeros((32, 32), dtype=np.float64)
     for i in range(line_array.shape[0]):
         LiMat = get_left_gmt_matrix(line_array[i, :])
-        tmat = np.matmul(np.matmul(LiMat, mask_2minus4), LiMat)
+        tmat = (LiMat@mask_2minus4)@LiMat
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:628]: nit: 
```suggestion
        tmat = (LiMat @ mask_2minus4) @ LiMat
```

File: clifford/tools/g3c/__init__.py [hunk 10]
```diff
@@ -560,14 +625,14 @@ def val_truncated_get_line_reflection_matrix(line_array, n_power):
...
         mat2solve += tmat
-    mat_val = np.matmul(mask1, mat2solve)
+    mat_val = mask1@mat2solve
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:630]:
```suggestion
    mat_val = mask1 @ mat2solve
```

File: clifford/tools/g3c/__init__.py [hunk 11]
```diff
@@ -560,14 +625,14 @@ def val_truncated_get_line_reflection_matrix(line_array, n_power):
...
     mat_val = mat_val[1:6, 1:6].copy()/line_array.shape[0]
     if n_power != 1:
         c_pow = 1
         while c_pow < n_power:
-            mat_val = np.matmul(mat_val, mat_val)
+            mat_val = mat_val@mat_val
             c_pow = c_pow*2
     return mat_val
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:637]: Would be a good use case for `np.linalg.matrix_power`, but I guess numba doesn't support it.

File: clifford/tools/g3c/cost_functions.py [hunk 12]
```diff
@@ -74,7 +74,7 @@ def midpoint_and_error_of_line_cluster_svd(line_cluster):
     Uses the SVD and explicit derivatives
     """
     line_cluster_array = np.array([l.value for l in line_cluster])
-    mat2solve = get_line_reflection_matrix(line_cluster)
+    mat2solve = get_line_reflection_matrix(line_cluster, 1)
```
Comment (eric-wieser) [inline clifford/tools/g3c/cost_functions.py:77]: Any point in this change?
```suggestion
    mat2solve = get_line_reflection_matrix(line_cluster)
```

File: clifford/tools/g3c/__init__.py [hunk 13]
```diff
@@ -428,10 +431,74 @@ def closest_points_on_circles(C1, C2, niterations=20):
...
+        # Project the centre of the circle onto the line
+        X = (A|Lpln)*einf*(A|Lpln)
+        if sphere_in_sphere(X*I5, bound_sphere):
+            # The circle and line intersect
+            PP = meet(Lpln, bound_sphere)
+            return point_pair_to_end_points(PP)[0]
+        else:
+            L2 = A ^ X ^ einf
+            PP = meet(L2, bound_sphere)
+            return point_pair_to_end_points(PP)[1]
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:464]: Codecov says these lines are not tested. Can you construct a test by hand that hits here, by making the plane and circle coplanar?

File: clifford/tools/g3c/__init__.py [hunk 14]
```diff
@@ -428,10 +431,74 @@ def closest_points_on_circles(C1, C2, niterations=20):
...
+            # Line passes through the centre of the circle
+            if abs((P | Y)[0]) < 1E-6:
+                # Line is perpendicular to plane of the circle
+                L2 = (A^project_points_to_circle([random_conformal_point()], C)[0]^einf).normal()
+            else:
+                # Just project the line
+                L2 = (L.normal() + (phi*L*phi)(3).normal()).normal()
+        elif abs((P | Y)[0]) < 1E-6:
+            # Line is perpendicular to the plane of the circle
+            L2 = A ^ Y ^ einf
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:483]: These lines are also not tested, a single test-case for each would be good.

File: clifford/tools/g3c/__init__.py [hunk 15]
```diff
@@ -281,17 +281,19 @@ def sphere_beyond_plane(sphere, plane):
     Check if the sphere is fully beyond the plane in the direction of
     the plane normal
     """
-    snorm = unsign_sphere(sphere)
-    return (snorm|plane)[0] < -get_radius_from_sphere(snorm)
+    restrue = ((meet(sphere, plane) ** 2)[0] < 0)
+    restrue = restrue and point_beyond_plane(normalise_n_minus_1((sphere * einf * sphere)(1)), plane)
+    return restrue
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:286]: nit:
```suggestion
    no_intersection = ((meet(sphere, plane) ** 2)[0] < 0)
    return no_intersection and point_beyond_plane(normalise_n_minus_1((sphere * einf * sphere)(1)), plane)
```

File: clifford/test/test_g3c_tools.py [hunk 16]
```diff
@@ -334,11 +371,135 @@ def test_enclosing_spheres(self):
...
+            X1, X2 = iterative_closest_points_circle_line(C, L, niterations=50)
+            X1Andreas = closest_point_on_circle_from_line(C, L)
+            X2Andreas = closest_point_on_line_from_circle(C, L)
+
+            assert_allclose((X1 ^ C).value, 0)
+            assert_allclose((X1Andreas ^ C).value, 0)
+            assert_allclose((X2 ^ L).value, 0)
+            assert_allclose((X2Andreas ^ L).value, 0)
+
+        # Choose explicit cases to ensure test coverage
```
Comment (eric-wieser) [inline clifford/test/test_g3c_tools.py:402]:
```suggestion
def test_closest_points_circle_line_corner_cases(self):
        # Choose explicit cases to ensure test coverage
```

File: clifford/tools/g3c/__init__.py [hunk 17]
```diff
@@ -428,10 +429,73 @@ def closest_points_on_circles(C1, C2, niterations=20):
...
+
+def iterative_closest_points_circle_line(C, L, niterations=20):
     """
     Given a circle C and line L this calculates the closest
-    points on each of them to the other
+    points on each of them to the other.
+
+    This is an iterative algorithm based on heuristics
+    Nonetheless it appears to give results on par with
+    closest_point_on_circle_from_line
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:498]:
```suggestion
    :func:`closest_point_on_circle_from_line`.
```
to get a link

File: clifford/tools/g3c/__init__.py [hunk 18]
```diff
@@ -560,16 +617,11 @@ def val_truncated_get_line_reflection_matrix(line_array, n_power):
     mat2solve = np.zeros((32, 32), dtype=np.float64)
```
Comment (eric-wieser) [inline clifford/tools/g3c/__init__.py:617]: On the line above this line, `n_power should be 1 or a power of 2` is nonsense. Maybe just add a type annotations, `n_power: int)`?

Comment (hugohadfield) [reply clifford/tools/g3c/__init__.py:616]: good spot

**Comment by eric-wieser** [comment]
Let's not merge until we have passing CI.

Also probably best to squash merge, this history is a mess.

**Pull request merged**